### PR TITLE
Adds reset password

### PR DESCRIPTION
**Pull request by Unraveler**
Adds logic to able password reset for a user.

Please review @terrestris/devs

### CODE REVIEW

File: shogun-lib/src/main/java/de/terrestris/shoguncore/controller/UserController.java [hunk 1]
```diff
@@ -167,4 +181,44 @@ public void changePassword(@RequestBody PasswordChange passwordChangeBody) {
             );
         }
     }
+
+    @PostMapping(value = "/resetPassword")
+    public void resetPassword(@RequestBody PasswordReset passwordResetBody, HttpServletRequest request){
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/controller/UserController.java:186]:
```suggestion
    public void resetPassword(@RequestBody PasswordReset passwordResetBody, HttpServletRequest request) {
```

File: shogun-lib/src/main/java/de/terrestris/shoguncore/controller/UserController.java [hunk 2]
```diff
@@ -167,4 +181,44 @@ public void changePassword(@RequestBody PasswordChange passwordChangeBody) {
...
+            URI appUrl = HttpUtil.getApplicationURIFromRequest(request);
+            OnPasswordResetRequestEvent event = new OnPasswordResetRequestEvent(user, request.getLocale(), appUrl.toString());
+            eventPublisher.publishEvent(event);
+        } catch (MailException | URISyntaxException e) {
+            LOG.error("Error sending mail: {}", e.getMessage());
+            LOG.trace("Full stack trace: ", e);
+
+            throw new ResponseStatusException(
+                HttpStatus.BAD_REQUEST,
+                messageSource.getMessage("registration.error.mail", null, request.getLocale())
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/controller/UserController.java:201]: Please set a custom message for this use case (e.g. `passwordReset.error.mail`).

File: shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        final SimpleMailMessage email = new SimpleMailMessage();
+        email.setTo(recipientAddress);
+        email.setSubject(subject);
+        email.setText(message);
+        email.setFrom(Objects.requireNonNull(env.getProperty("support.email"), "Environment variable support.email not set"));
+
+        return email;
+    }
+
+
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java:97]:
```suggestion
```

File: shogun-lib/src/main/java/de/terrestris/shoguncore/service/UserService.java [hunk 4]
```diff
@@ -224,6 +247,15 @@ private boolean emailExists(final String email) {
         return this.repository.findOne(UserSpecification.findByMail(email)).isPresent();
     }
 
+    /**
+     * Find user by giving email
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/service/UserService.java:251]:
```suggestion
     * Find an user by the given email.
```

File: shogun-lib/src/main/java/de/terrestris/shoguncore/controller/UserController.java [hunk 5]
```diff
@@ -167,4 +181,44 @@ public void changePassword(@RequestBody PasswordChange passwordChangeBody) {
...
+            );
+        }
+    }
+
+    @GetMapping(value = "/password/resetPassword/confirm")
+    public RedirectView confirmPasswordReset(@RequestParam("token") String token, HttpServletRequest request) throws URISyntaxException {
+        RedirectView redirectView = new RedirectView();
+        try {
+            final String result = service.validateVerificationToken(token, request.getLocale(), "passwordReset");
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/controller/UserController.java:210]: Please introduce an enum for the module strings (here `passwordReset`).

File: shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        this.handleConfirmPasswordReset(event);
+    }
+
+    private void handleConfirmPasswordReset(final OnPasswordResetRequestEvent event) {
+        final User user = event.getUser();
+        final String token = UUID.randomUUID().toString();
+
+        userService.createVerificationTokenForUser(user, token);
+
+        LOG.info("Generated verification token for user " + user.getUsername());
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java:51]:
```suggestion
        LOG.info("Generated verification token for user {}", user.getUsername());
```

File: shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+        try {
+            final SimpleMailMessage email = constructPasswordResetEmailMessage(event.getAppUrl(), user, token, event.getLocale());
+            mailSender.send(email);
+        } catch (Exception e) {
+            throw new MailException(e.getMessage());
+        }
+
+        LOG.info("Successfully sent password reset mail to user " + user.getEmail());
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java:60]:
```suggestion
        LOG.info("Successfully sent password reset mail to user {}", user.getEmail());
```

File: shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    /**
+     * Constructs a Reset password Confirm Email based on the users' properties (name, language, etc.)
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shoguncore/listener/PasswordResetListener.java:64]:
```suggestion
     * Constructs a reset password confirm email based on the users' properties (name, language, etc.)
```

**Comment by Unraveler** [comment]
@dnlkoch you can check it again

**Review by dnlkoch — APPROVED** [review]

**Pull request merged**